File: scripts/import-hotels.py
import asyncio
import logging
from pathlib import Path
from prisma import Prisma
from prisma.errors import UniqueViolationError

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

async def clean_database(prisma: Prisma):
    """Deletes all data from Room, Availability, and Hotel tables."""
    try:
        logger.info('Cleaning database...')
        # Delete in reverse order of dependency
        await prisma.room.delete_many()
        logger.info('Deleted rooms')
        await prisma.availability.delete_many()
        logger.info('Deleted availabilities')
        await prisma.hotel.delete_many()
        logger.info('Deleted hotels')
        logger.info('Database cleaned successfully')
    except Exception as e:
        logger.error(f'Error cleaning database: {e}')
        raise # Re-throw the error to stop the import process

async def import_hotels(prisma: Prisma):
    """Imports hotel data from hotel_details.txt into the database."""
    hotels_processed = 0
    hotels_created = 0
    hotels_skipped = 0 # Counter for skipped duplicates


    logger.info('Starting hotel import...')

    # Clean the database first (optional, uncomment to enable)
    # await clean_database(prisma)

    file_path = Path.cwd() / 'hotel_details.txt'
    logger.info(f'Reading file from: {file_path}')

    if not file_path.exists():
        logger.error(f'Error: File not found at {file_path}')
        return # Exit the function gracefully

    current_hotel = {}
    line_count = 0

    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if len(line) == 0:
                continue
            if line[0] == 'N':
                current_hotel['name'] = line[6:]
            elif line[0] == 'U':
                current_hotel['url'] = line[5:]
            elif line[0] == 'L':
                current_hotel['location'] = line[10:]
            elif line[0] == 'R':
                current_hotel['rating'] = line[8:]
            elif line[0] == '=':
                hotels_processed += 1
                if current_hotel.get('name') and current_hotel.get('url'):
                    rating_str = current_hotel.get('rating')
                    rating = None
                    if rating_str:
                        try:
                            rating = float(rating_str)
                        except ValueError:
                            logger.warning(f"Invalid rating format '{rating_str}' for hotel {current_hotel.get('name')}. Setting rating to null.")

                    hotel_data = {
                        'name': current_hotel['name'],
                        'url': current_hotel['url'],
                        'location': current_hotel.get('location'),
                        'rating': rating
                    }
                    
                    
                    try:
                        await prisma.hotel.create(data=hotel_data)
                        hotels_created += 1
                    except UniqueViolationError:
                        logger.warning(f"Skipping duplicate URL found in file: {hotel_data['url']} (Hotel: {hotel_data['name']})")
                        hotels_skipped += 1
    


async def main():
    """Main function to connect to DB, run import, and disconnect."""
    prisma = Prisma(log_queries=False) # Optionally log SQL queries
    """"""
    try:
        await prisma.connect()
        logger.info('Database connection established.')
        await import_hotels(prisma)
        #await clean_database(prisma)
    except Exception as e:
        logger.error(f"An error occurred in main execution: {e}")
    finally:
        if prisma.is_connected():
            await prisma.disconnect()
            logger.info('Database connection closed.')

if __name__ == '__main__':
    asyncio.run(main())

File: scripts/import-hotels.js
import { PrismaClient } from '@prisma/client';
import fs from 'fs';
import readline from 'readline';
import path from 'path';

const prisma = new PrismaClient();

async function cleanDatabase() {
  try {
    console.log('Cleaning database...');
    // Delete in reverse order of dependency
    await prisma.room.deleteMany();
    console.log('Deleted rooms');
    await prisma.availability.deleteMany();
    console.log('Deleted availabilities');
    await prisma.hotel.deleteMany();
    console.log('Deleted hotels');
    console.log('Database cleaned successfully');
  } catch (error) {
    console.error('Error cleaning database:', error);
    throw error; // Re-throw the error to stop the import process
  }
}

async function importHotels() {
  let hotelsProcessed = 0;
  let hotelsCreated = 0;
  let hotelsSkipped = 0; // Counter for skipped duplicates
  
  try {
    console.log('Starting hotel import...');
    
    // Clean the database first
    //await cleanDatabase();
    
    const filePath = path.resolve(process.cwd(), 'hotel_details.txt');
    console.log(`Reading file from: ${filePath}`);
    
    if (!fs.existsSync(filePath)) {
        console.error(`Error: File not found at ${filePath}`);
        process.exit(1);
    }

    const fileStream = fs.createReadStream(filePath);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity
    });

    let currentHotel = {};
    let lineCount = 0;

    for await (const line of rl) {
      lineCount++;
      if (line.trim() === '===') {
        hotelsProcessed++;
        // Process the completed hotel
        if (currentHotel.name && currentHotel.url) {
            const hotelData = {
                name: currentHotel.name,
                url: currentHotel.url,
                location: currentHotel.location || null,
                rating: currentHotel.rating ? parseFloat(currentHotel.rating) : null
            };

          try {
            // console.log(`Attempting to create hotel: ${hotelData.name} (URL: ${hotelData.url})`); // Uncomment for detailed logging
            await prisma.hotel.create({
              data: hotelData
            });
            hotelsCreated++;
            
            // Log progress every 100 hotels CREATED
            if (hotelsCreated > 0 && hotelsCreated % 100 === 0) {
              console.log(`Processed ${hotelsProcessed} sections, Created ${hotelsCreated} hotels, Skipped ${hotelsSkipped} duplicates`);
            }
          } catch (error) {
            if (error.code === 'P2002') { // Prisma unique constraint violation code
              console.warn(`Skipping duplicate URL found in file: ${hotelData.url} (Hotel: ${hotelData.name})`);
              hotelsSkipped++;
            } else {
              console.error(`Error creating hotel: ${hotelData.name} (URL: ${hotelData.url}) at line ~${lineCount}`, error);
              // Optionally decide if you want to stop on other errors
              // process.exit(1);
            }
          }
        } else {
            console.warn(`Skipping incomplete hotel data at line ~${lineCount}:`, currentHotel);
        }
        currentHotel = {}; // Reset for the next hotel
      } else {
        const parts = line.split(': ');
        if (parts.length >= 2) {
            const key = parts[0].trim().toLowerCase();
            const value = parts.slice(1).join(': ').trim(); // Handle cases where value might contain ':'
            if (key && value) {
              currentHotel[key] = value;
            }
        } else if (line.trim()) {
            // Handle potential malformed lines if necessary
            // console.warn(`Malformed line ${lineCount}: ${line.trim()}`);
        }
      }
    }

    // Process the last hotel if file doesn't end with ===
    if (Object.keys(currentHotel).length > 0 && currentHotel.name && currentHotel.url) {
        hotelsProcessed++;
        const hotelData = {
            name: currentHotel.name,
            url: currentHotel.url,
            location: currentHotel.location || null,
            rating: currentHotel.rating ? parseFloat(currentHotel.rating) : null
        };
        try {
            await prisma.hotel.create({ data: hotelData });
            hotelsCreated++;
        } catch (error) {
            if (error.code === 'P2002') {
                console.warn(`Skipping duplicate URL found in file (last hotel): ${hotelData.url} (Hotel: ${hotelData.name})`);
                hotelsSkipped++;
            } else {
                console.error(`Error creating last hotel: ${hotelData.name} (URL: ${hotelData.url})`, error);
            }
        }
    }

    console.log('-------------------------------');
    console.log('Import completed!');
    console.log(`Total sections processed in file: ${hotelsProcessed}`);
    console.log(`Hotels successfully created: ${hotelsCreated}`);
    console.log(`Duplicate URLs skipped: ${hotelsSkipped}`);
    console.log('-------------------------------');

  } catch (error) {
    // Catch errors from cleanDatabase or file reading
    console.error('Fatal error during import process:', error);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
    console.log('Database connection closed.');
  }
}

importHotels(); 